Source/Dafny/Resolver.cs

```diff
@@ -18571,10 +18571,21 @@ public class SubsetConstraintGhostChecker : ProgramTraverser {
       return skip;
     }
 
+    private bool IsFieldSpecification(string field, object parent) {
+      return field != null && parent != null && (
+        (parent is Statement && field == "SpecificationSubExpressions") ||
+        (parent is Function && (field is "Req.E" or "Reads.E" or "Ens.E" or "Decreases.Expressions")) ||
+        (parent is Method && (field is "Req.E" or "Mod.E" or "Ens.E" or "Decreases.Expressions"))
+      );
+    }
+
     public override bool Traverse(Expression expr, [CanBeNull] string field, [CanBeNull] object parent) {
       if (expr == null) {
         return false;
       }
+      if (IsFieldSpecification(field, parent)) {
```

Comment:
I would expect us to have all error messages of the form "...is not compilable..." to be created after a de-ghosting phase in the compiler pipeline, after which all Expressions should be non-ghost, so I'm a bit surprised we need the change this PR makes but I guess we need it with the current architecture.

Reply:
That seems like a reasonable expectation. I would go one step further: Everything has to happen before compilation, and there should be no such "is not compilable" error thrown in practice by the compiler.
Indeed, ideally subset types should be verified and then removed at compile time, and I'm currently working on that.